examples/vision/siamese.py

```diff
@@ -0,0 +1,247 @@
+        labels += [1]
+    
+        # add a not matching example
+        label2 = random.randint(0, num_classes-1)
+        while label2 == label1:
+            label2 = random.randint(0, num_classes-1)
+
+        idx2 = random.choice(digit_indices[label2])
+        x2 = x[idx2]
+        
+        pairs += [[x1, x2]]
+        labels += [0]
+
+    return np.array(pairs), np.array(labels)
+
+pairs_train, labels_train = make_pairs(x_train, y_train)
+pairs_test, labels_test = make_pairs(x_test, y_test)
+
+"""remember that our labels are still in int data type, so convert them to float"""
+
+labels_train = labels_train.astype('float32')
+labels_test = labels_test.astype('float32')
+
+"""
+## Create dataset 
+"""
+
+#  0 index of pairs_train contain one image of pair and at index 1 the other image of pair is present
+#  we can split entire vector of pairs into two havles, later we can generate tf.data.Dataset using these halves
+# we will do the same for labels
+
+x1 = pairs_train[:,0]
+x2 = pairs_train[:,1]
+# x1.shape = (120000, 28, 28)
+
+y1, y2 = pairs_test[:,0],pairs_test[:,1]
+
+train_pair = tf.data.Dataset.from_tensor_slices((x1, x2))
+train_label = tf.data.Dataset.from_tensor_slices(labels_train)
+train_ds = tf.data.Dataset.zip((train_pair, train_label)).batch(16)
+
+test_pair = tf.data.Dataset.from_tensor_slices((y1, y2))
+test_label = tf.data.Dataset.from_tensor_slices(labels_test)
+test_ds = tf.data.Dataset.zip((test_pair, test_label)).batch(16)
+
+"""
+## Visualize
+"""
+
+to_show = 10
+num_col = 5
+num_row = to_show//num_col
+# plot images
+fig, axes = plt.subplots(num_row, num_col, figsize=(1.5*num_col,2*num_row))
+for images, labels in train_ds.take(1):
+  for i in range(to_show):
+      ax = axes[i//num_col, i%num_col]
+      # images[0][i][:,:,0] -> because it is (28,28,1) and imshow takes (28,82)
+      # ax.imshow(tf.concat([images[0][i][:,:,0],images[1][i][:,:,0]],axis=1))
+
+      ax.imshow(tf.concat([images[0][i],images[1][i]],axis=1))
+
+      ax.set_title('Label: {}'.format(labels[i]))
+plt.tight_layout()
+plt.show()
+
+"""
+## Model
+
+We can make model as described in diagram in the start. The distance metric we will be using for finding distance between embedding vectors is euclidean_distance. We will be having two input layers, each of these two lead to a dense network which results in embedding vector of having 10 rows. Lambda layer will take each of these two embedding vectors and apply euclidean distance function, the output is passed from one Dense layer later.
+"""
+
+def euclidean_distance(vects):
+    x, y = vects
+    sum_square = K.sum(K.square(x - y), axis=1, keepdims=True)
+    return K.sqrt(K.maximum(sum_square, K.epsilon()))
+
+
+
+input = Input((28,28,1))
+x = Conv2D(4, (5,5), activation = 'tanh')(input)
+x = AveragePooling2D(pool_size = (2,2))(x)
+x = Conv2D(16, (5,5), activation = 'tanh')(x)
+x = AveragePooling2D(pool_size = (2,2))(x)
+x = Flatten()(x)
+
+tf.keras.layers.BatchNormalization()
```

Comment:
This layer is not connected to anything.

Reply:
Thanks, really sorry for I confused wrong .py file with correct one while commiting, resolved now.